File: problems/problems_55/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"


using namespace std;
using json = nlohmann::json;

class Solution {
public:
    bool canJump(vector<int>& nums) {
		int max_dis = 0, n = static_cast<int>(nums.size());
		for (int i = 0; i < n; i++) {
			max_dis = max(max_dis, i + nums[i]);
			if (max_dis >= n - 1) {
				return true;
			}
			if (i >= max_dis) {
				return false;
			}
		}
		return false;
    }
};

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	Solution solution;
	vector<int> nums = json::parse(inputArray.at(0));
	return solution.canJump(nums);
}

File: problems/problems_55/solution.rs
use serde_json::{json, Value};

pub struct Solution;

impl Solution {
    pub fn can_jump(nums: Vec<i32>) -> bool {
		let mut max_dis: i32 = 0;
		for i in 0..nums.len() {
			max_dis = max_dis.max(i as i32 + nums[i]);
			if max_dis >= nums.len() as i32 - 1 {
				return true;
			}
			if i as i32 >= max_dis {
				return false;
			}
		}
		false
    }
}

#[cfg(feature = "solution_55")]
pub fn solve(input_string: String) -> Value {
	let input_values: Vec<String> = input_string.split('\n').map(|x| x.to_string()).collect();
	let nums: Vec<i32> = serde_json::from_str(&input_values[0]).expect("Failed to parse input");
	json!(Solution::can_jump(nums))
}
